subxt/src/backend/mod.rs

```diff
@@ -392,121 +434,212 @@ mod test {
                 }
             }
 
-            fn pop(&mut self, key: Vec<u8>) -> Item {
+            pub fn pop(&mut self, key: Vec<u8>) -> Message<Item> {
                 self.items.get_mut(&key).unwrap().pop_front().unwrap()
             }
         }
 
-        struct Subscription {
-            sender: mpsc::Sender<RpcResult<Vec<Item>>>,
-            receiver: mpsc::Receiver<RpcResult<Vec<Item>>>,
+        pub struct Subscription {
+            sender: mpsc::Sender<Item>,
         }
 
         impl Subscription {
-            fn new() -> Self {
+            pub fn new() -> (Self, mpsc::Receiver<Item>) {
                 let (sender, receiver) = mpsc::channel(32);
-                Self { sender, receiver }
+                (Self { sender }, receiver)
             }
 
-            async fn from_iter<
-                T: Serialize,
-                S: IntoIterator<Item = RpcResult<Vec<RpcResult<T>>>>,
-            >(
-                items: S,
-            ) -> Self {
-                let sub = Self::new();
-                for i in items {
-                    let i: RpcResult<Vec<Item>> = i.map(|items| {
-                        items
-                            .into_iter()
-                            .map(|item| item.map(|i| serde_json::to_string(&i).unwrap()))
-                            .collect()
-                    });
-                    sub.write(i).await
-                }
-                sub
+            pub async fn write(&self, items: Message<Item>) {
+                match items {
+                    Message::Many(items) => {
+                        for i in items.unwrap() {
+                            self.sender.send(i).await.unwrap()
+                        }
+                    }
+                    Message::Single(item) => self.sender.send(item).await.unwrap(),
+                };
             }
 
-            async fn read(&mut self) -> RpcResult<Vec<Item>> {
-                self.receiver.recv().await.unwrap()
-            }
+            pub async fn write_delayed(&self, items: Message<Item>) {
+                let sender = self.sender.clone();
+                tokio::spawn(async move {
+                    tokio::time::sleep(Duration::from_millis(500)).await;
 
-            async fn write(&self, items: RpcResult<Vec<Item>>) {
-                self.sender.send(items).await.unwrap()
+                    match items {
+                        Message::Many(items) => {
+                            for i in items.unwrap() {
+                                let _ = sender.send(i).await;
+                            }
+                        }
+                        Message::Single(item) => sender.send(item).await.unwrap(),
+                    };
+                });
             }
         }
 
         struct Data {
-            request: MockDataTable,
-            subscription: Subscription,
+            data_table: MockDataTable,
+            subscription_channel: Option<Subscription>,
```

Comment:
can you move the subscription_channel to the subscription handler? or is this needed to keep the channel alive?

Reply:
yup, sometimes we reference it inside request handlers